### PR TITLE
[Part 5] Set actor died error message in ActorDiedError

### PR DESCRIPTION
**Pull request by rkooo567**
## Why are these changes needed?

This is the second last PR to improve `ActorDiedError` exception. 

This propagates Actor death cause metadata to the ray error object. In this way, we can raise a better actor died error exception.

After this PR is merged, I will add more metadata to each error message and write a documentation that explains when each error happens. 

TODO
- [x] Fix test failures
- [x] Add unit tests
- [x] Fix Java/cpp cases

Follow up PRs
- Not allowing nullptr for RayErrorInfo input.

## Related issue number

https://github.com/ray-project/ray/issues/20534

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/exceptions.py [hunk 1]
```diff
@@ -243,7 +246,27 @@ def __str__(self):
             return ("The actor died because of an error" +
                     " raised in its creation task, " +
                     self.creation_task_error.__str__())
-        return ("The actor died unexpectedly before finishing this task.")
+
+        base_error_msg = (
+            "The actor died unexpectedly before finishing this task")
+        if self.cause:
+            if self.cause.HasField("worker_died_context"):
```
Comment (rkooo567) [inline python/ray/exceptions.py:253]: I think alternative (probably the cleaner approach) approach is to have a different exception type per each exception (all of them inherited from ActorDiedError).

And then we can write a documentation that explains each of failure.

Comment (rkooo567) [reply python/ray/exceptions.py:253]: Any thought on this?

File: src/ray/core_worker/transport/direct_actor_task_submitter.cc [hunk 2]
```diff
@@ -110,19 +110,21 @@ Status CoreWorkerDirectActorTaskSubmitter::SubmitTask(TaskSpecification task_spe
...
     {
       absl::MutexLock lock(&mu_);
-      auto queue = client_queues_.find(task_spec.ActorId());
-      auto &death_cause = queue->second.death_cause;
-      error_type = GenErrorTypeFromDeathCause(death_cause.get());
-      error_info = GetErrorInfoFromActorDeathCause(death_cause.get());
+      const auto queue = client_queues_.find(task_spec.ActorId());
+      const std::unique_ptr<rpc::ActorDeathCause> &death_cause =
+          queue->second.death_cause;
+      RAY_CHECK(death_cause != nullptr);
```
Comment (scv119) [inline src/ray/core_worker/transport/direct_actor_task_submitter.cc:119]: hmm in general we shouldn't crash if the death_cause is unknown?

File: src/ray/protobuf/common.proto [hunk 3]
```diff
@@ -170,6 +171,57 @@ message RayException {
   string formatted_exception_string = 3;
 }
 
+message ActorDeathCause {
```
Comment (scv119) [inline src/ray/protobuf/common.proto:174]: nit: enum is a bit overkill here?

File: python/ray/exceptions.py (hunk 1 above)
Comment (scv119) [reply python/ray/exceptions.py:253]: if we remove enum it should be simpler...

File: python/ray/exceptions.py [hunk 4]
```diff
@@ -211,7 +211,8 @@ class RayActorError(RayError):
     executing a task, or because a task is submitted to a dead actor.
 
     If the actor is dead because of an exception thrown in its creation tasks,
-    RayActorError will contains this exception.
+    RayActorError will contain creation_task_error, which is used to
```
Comment (ericl) [inline python/ray/exceptions.py:214]:
```suggestion
    RayActorError will contain the creation_task_error, which is used to
```

File: python/ray/exceptions.py [hunk 5]
```diff
@@ -243,7 +246,27 @@ def __str__(self):
...
+                        f"{self.cause.node_died_context.error_message}")
+            elif self.cause.HasField("owner_died_context"):
+                return (f"{base_error_msg} "
+                        f"{self.cause.owner_died_context.error_message}")
+            elif self.cause.HasField("killed_by_app_context"):
+                return (f"{base_error_msg} "
+                        f"{self.cause.killed_by_app_context.error_message}")
+            elif self.cause.HasField("out_of_scope_context"):
+                return (f"{base_error_msg} "
+                        f"{self.cause.out_of_scope_context.error_message}")
```
Comment (ericl) [inline python/ray/exceptions.py:267]: Should the else case assert false?

File: python/ray/exceptions.py (hunk 1 above)
Comment (ericl) [reply python/ray/exceptions.py:253]: Yeah maybe we remove the enum? Seems like a lot of code complexity when we could have instead stored the message as {"error_class": "WorkerDied", "message": the actual message}.

File: src/ray/core_worker/task_manager.h [hunk 6]
```diff
@@ -140,6 +140,8 @@ class TaskManager : public TaskFinisherInterface, public TaskResubmissionInterfa
   /// \param[in] error_type The type of the specific error.
   /// \param[in] status Optional status message.
   /// \param[in] ray_error_info The error information of a given error type.
+  /// Nullptr means that there's no error information.
+  /// TODO(sang): Remove nullptr case. Every error message should have metadata.
```
Comment (ericl) [inline src/ray/core_worker/task_manager.h:144]: How about we remove nullptr right away and instead add "TODO / unknown cause" in the call sites? Supporting special case missing information has bitten us in the past (e.g., raylet pid gets lost).

File: src/ray/core_worker/transport/direct_actor_task_submitter.cc (hunk 2 above)
Comment (ericl) [reply src/ray/core_worker/transport/direct_actor_task_submitter.cc:119]: I think it would be better to just make sure it's non-nullable (e.g., pass by ref). The call site shouldn't accept null causes.

File: python/ray/exceptions.py (hunk 1 above)
Comment (rkooo567) [reply python/ray/exceptions.py:253]: It is actually not enum, but oneof. 

I made it have oneof because it will have more metadata per error type (that might differ) in the future. For example, we probably want to include why workers are dead when actor is dead by worker died. 

But alternatively, I can just construct the complete error message from GCS and not include metadata at all. That will certainly simplify the code here. (and in this code, we just print error_message field). Let me try with this approach.

Also, what do you guys think about exception types? Do you prefer to have 1 exception ActorDiedError, or `class ActorOutOfScope(ActorDied)` error?

File: src/ray/core_worker/transport/direct_actor_task_submitter.cc (hunk 2 above)
Comment (rkooo567) [reply src/ray/core_worker/transport/direct_actor_task_submitter.cc:119]: This requires some refactoring. Right now, this death_cause is for some reasons within the struct, and this has invariant, "if actor is dead this value must be set". What about doing refactoring in a follow up PR?

File: src/ray/protobuf/common.proto (hunk 3 above)
Comment (rkooo567) [reply src/ray/protobuf/common.proto:174]: As mentioned above, this is to add additional metadata per error type. But alternatively, I can just construct the full error message from GCS and make it only have one field (error_message). Which approach do you prefer?

**Comment by rkooo567** [comment]
So, some of requests make sense, but it'd make PR very big.

There are 2 points;

1. Should we use oneof or just a single ActorDied death cause with "error message". The motivation of oneof is to allow to specify metadata per error type (e.g., metadata required for WorkerDied death cause might be different from OwnerDied). We can remove oneof if we completely compose the error message from the GCS. I will go with this approach, but I'd prefer to do it in the next PR
2. Do not allow nullptr. This is a good suggestion, but MarkTaskObjectFailed is used for not just actor died error, and not supporting nullptr requires big changes. I can do it in a follow up. 

TL;DR. I will make 2 follow up PRs to remove nullptr input type + remove oneof and improve actor died error metadata after this PR.

File: src/ray/common/ray_object.cc [hunk 7]
```diff
@@ -96,12 +95,7 @@ RayObject::RayObject(rpc::ErrorType error_type, const rpc::RayErrorInfo *ray_err
     return;
   }
 
-  RAY_CHECK(ray_error_info->has_actor_init_failure());
-  // This is temporarily here because changing this requires changes in all language
-  // frontend.
-  // TODO(sang, lixin): Remove it.
-  const auto error_buffer =
-      MakeSerializeErrorBuffer<rpc::RayException>(ray_error_info->actor_init_failure());
+  const auto error_buffer = MakeSerializedErrorBuffer<rpc::RayErrorInfo>(*ray_error_info);
```
Comment (rkooo567) [inline src/ray/common/ray_object.cc:98]: @lixin-wei can you help fixing the Java case? Java side will be incompatible to this change.

Comment (lixin-wei) [reply src/ray/common/ray_object.cc:98]: The code has changed a lot since my last modification. I read your discussion but didn't get the context. Would you please say more about what happened, why it will affect java cases, and what should I do?

Comment (rkooo567) [reply src/ray/common/ray_object.cc:98]: Previously, all of ray error object didn't have body, except the actor creation exception. Now, we want to allow to pass protobuf-defined body so that we can write a better error message on failures. 

Currently, we allow to specify body for all ray objects. And by default we changed this to `message RayErrorInfo `, and `RayErrorInfo ` contains `RayException` (which was for actor creation exception). This is incompatible to Java / Cpp (since they assume if there's any body on ray object, it is `RayException`, but it is `RayErrorInfo` now).

Comment (rkooo567) [reply src/ray/common/ray_object.cc:98]: You can see `serialization.py`

File: src/ray/protobuf/common.proto [hunk 8]
```diff
@@ -182,6 +182,7 @@ enum ErrorType {
 message RayErrorInfo {
   oneof error {
     RayException actor_init_failure = 1;
+    ActorDeathCause actor_died_error = 2;
```
Comment (lixin-wei) [inline src/ray/protobuf/common.proto:185]: There is already an `CreationTaskFailureContext` in `ActorDeathCause`, I think we can remove the `actor_init_failure` here. Do you think so?

**Review by lixin-wei — CHANGES_REQUESTED** [review]

File: src/ray/protobuf/common.proto (hunk 8 above)
Comment (rkooo567) [reply src/ray/protobuf/common.proto:185]: Hmm yeah good point. Let me fix that.

File: python/ray/serialization.py [hunk 9]
```diff
@@ -218,12 +233,8 @@ def _deserialize_object(self, data, metadata, object_ref):
...
                 return WorkerCrashedError()
             elif error_type == ErrorType.Value("ACTOR_DIED"):
-                if data:
-                    pb_bytes = self._deserialize_msgpack_data(
-                        data, metadata_fields)
-                    if pb_bytes:
-                        return RayError.from_bytes(pb_bytes)
-                return RayActorError()
+                return self._deserialize_actor_died_error(
+                    data, metadata_fields)
```
Comment (lixin-wei) [inline python/ray/serialization.py:237]: Can we remove error_type? Since we have a RayError PB schema here,  we can move error_type into RayError, instead of appending an meta field to the bytes. That will make code cleaner and more matainable.

**Review by lixin-wei — DISMISSED** [review]

File: src/ray/common/ray_object.cc (hunk 7 above)
Comment (lixin-wei) [reply src/ray/common/ray_object.cc:98]: I see, thanks! But before that, I've some questions, commented.

File: python/ray/serialization.py (hunk 9 above)
Comment (rkooo567) [reply python/ray/serialization.py:237]: You mean `error_type == ErrorType.Value("ACTOR_DIED"):` -> this one?

It is too big scope to handle here. Consider creating an issue?

File: src/ray/common/ray_object.cc (hunk 7 above)
Comment (rkooo567) [reply src/ray/common/ray_object.cc:98]: Thanks! I have no Ray java setup on my machine haha..

File: python/ray/serialization.py (hunk 9 above)
Comment (rkooo567) [reply python/ray/serialization.py:237]: But agreed this could be cleaned up by being included within the protobuf

File: src/ray/common/ray_object.cc (hunk 7 above)
Comment (rkooo567) [reply src/ray/common/ray_object.cc:98]: Do you know who I should contact to for cpp API? I assume Cpp APi also needs some changes

Comment (rkooo567) [reply src/ray/common/ray_object.cc:98]: (though for some reasons the cpp tests don't seem to fail!)

Comment (lixin-wei) [reply src/ray/common/ray_object.cc:98]: Since this PR is large enough, I think maybe it's better to keep the compatibility in this PR, and open follow-up PRs to change Java/CPP deserialization. Do you think so? Then you won't be blocked by us.

You can write the `RayException` in `RayError. actor_died_error` to object store first to keep the compatibility.

Comment (lixin-wei) [reply src/ray/common/ray_object.cc:98]: For CPP, maybe you can ping @qicosmos

Comment (rkooo567) [reply src/ray/common/ray_object.cc:98]: > Since this PR is large enough, I think maybe it's better to keep the compatibility in this PR, and open follow-up PRs to change Python/Java/CPP deserialization

What do you mean by "keep compatibility" here? I am asking it because this is the PR that breaks the compatibility (all other PRs are already merged).

Comment (rkooo567) [reply src/ray/common/ray_object.cc:98]: What part of this PR do you think can be isolated in this case?

Comment (lixin-wei) [reply src/ray/common/ray_object.cc:98]: I meant writing RayException to object store first(But keep all refactor you made). And change it to write RayError in the next PR.

Anyway, I'm fine with both approaches. I think 50+ lines will be modified.

I asked @qicosmos, he said CPP worker doesn't support exception now. So you we can leave CPP worker as it is, I think.

Comment (rkooo567) [reply src/ray/common/ray_object.cc:98]: Hmm I see. So your suggestion is to only modify cpp side in this PR, and make another PR to actually change it to RayErrorInfo.

I am fine with that approach. Lmk what you prefer.

Comment (lixin-wei) [reply src/ray/common/ray_object.cc:98]: I prefer opening another PR. I'll open a PR to change both Python and Java after you merge this PR(for Python I'll copy the code you wrote in this PR). Do you think it's OK?

Comment (lixin-wei) [reply src/ray/common/ray_object.cc:98]: Oh nope, I think I prefer not opening another PR now. Since I realized that I need to copy and debug your python test.

I'll fix the Java test in this PR!

Comment (rkooo567) [reply src/ray/common/ray_object.cc:98]: Sounds good haha

File: src/ray/protobuf/common.proto (hunk 8 above)
Comment (rkooo567) [reply src/ray/protobuf/common.proto:184]: This is fixed

**Comment by rkooo567** [comment]
All tests except Java have fixed, and @lixin-wei will push changes to this PR.

**Comment by lixin-wei** [comment]
I'll do this in monday

File: python/ray/exceptions.py [hunk 10]
```diff
@@ -247,11 +262,32 @@ def get_creation_task_error(self):
...
+            return (f"{self.base_error_msg} "
+                    f"{cause.owner_died_context.error_message}")
+        elif cause.HasField("killed_by_app_context"):
+            return (f"{self.base_error_msg} "
+                    f"{cause.killed_by_app_context.error_message}")
+        elif cause.HasField("out_of_scope_context"):
+            return (f"{self.base_error_msg} "
+                    f"{cause.out_of_scope_context.error_message}")
+        else:
+            assert False, f"Unknown death cause. {cause}"
```
Comment (ericl) [inline python/ray/exceptions.py:290]: This still looks like unnecessary complexity here. Can we use a generalized schema so you can write:

```return "The actor died: {self.base_error_msg}: {self.cause}"?```

**Review by ericl — DISMISSED** [review]
I'd like to see the enum go away as it seems to be causing unnecessary code verbosity.

File: python/ray/exceptions.py (hunk 10 above)
Comment (rkooo567) [reply python/ray/exceptions.py:290]: So, I'd prefer this to be in a separate PR. What about I make a PR to generalize the schema (I am only going to have error_message and construct all metadata into it), merge that, and then we come back to this PR after?

File: src/ray/core_worker/task_manager.h (hunk 6 above)
Comment (rkooo567) [reply src/ray/core_worker/task_manager.h:149]: This will be handled in the next clean up PR!

**Comment by rkooo567** [comment]
Current error messages. 
```python3
The actor died unexpectedly before finishing this task.
	class_name: test_actor_failure_per_type.<locals>.Actor
	actor_id: 09870d4683d850e08923954401000000
	namespace: 6ba52822-fa0d-444b-b1f5-64d4582f928d
The actor is dead because it was killed by `ray.kill`.

The actor died unexpectedly before finishing this task.
	class_name: test_actor_failure_per_type.<locals>.Actor
	actor_id: 727a9d7bf50c0ce2968dc9cc01000000
	pid: 13255
	namespace: 6ba52822-fa0d-444b-b1f5-64d4582f928d
	ip: 127.0.0.1
(ip=127.0.0.1) The actor is dead because its worker process has died. Worker exit type: SYSTEM_ERROR_EXIT

The actor died unexpectedly before finishing this task.
	class_name: test_actor_failure_per_type.<locals>.Actor
	actor_id: 454a7f373217d77cbeaf839801000000
	namespace: 6ba52822-fa0d-444b-b1f5-64d4582f928d
The actor is dead because its owner has died. Owner Id: c7eb230a6c9e0bc8c0b696cf5b94f5264fa6b50c81a3142cbc0244e8 Owner Ip address: 127.0.0.1 Owner worker exit type: INTENDED_EXIT

The actor died unexpectedly before finishing this task.
	class_name: test_actor_failure_per_type.<locals>.Actor
	actor_id: bf35bf3c6de853b775cbb60a01000000
	pid: 13278
	namespace: 6ba52822-fa0d-444b-b1f5-64d4582f928d
	ip: 127.0.0.1
(ip=127.0.0.1) The actor is dead because its node has died. Node Id: 08118553b8c54a6c394c1fd867e05c2690a8bcde870293ea0d311035
```

**Comment by rkooo567** [comment]
@ericl @scv119 This is ready to review again.I will handle all test failures by today (should not be impacting reviewing the PR).

**Comment by rkooo567** [comment]
@lixin-wei any idea about java failure?

**Comment by lixin-wei** [comment]
Seems test is hanging in `CoreWorkerProcess::Shutdown()`, [download full stack dump](https://github.com/ray-project/ray/files/7897281/pstack.log)
```
Thread 2 (Thread 0x7f3cb7873700 (LWP 1115)):
--
  | #0  __pthread_clockjoin_ex (threadid=139896360511232, thread_return=0x0, clockid=<optimized out>, abstime=<optimized out>, block=<optimized out>) at pthread_join_common.c:145
  | #1  0x00007f3cb7ab5047 in std::thread::join() () from /lib/x86_64-linux-gnu/libstdc++.so.6
  | #2  0x00007f3c3b467984 in ray::core::CoreWorker::WaitForShutdown() () from /tmp/ray/1642573567986/libcore_worker_library_java.so
  | #3  0x00007f3c3b487e5a in ray::core::CoreWorkerProcessImpl::RemoveWorker(std::shared_ptr<ray::core::CoreWorker>) () from /tmp/ray/1642573567986/libcore_worker_library_java.so
  | #4  0x00007f3c3b488e45 in ray::core::CoreWorkerProcessImpl::ShutdownDriver() () from /tmp/ray/1642573567986/libcore_worker_library_java.so
  | #5  0x00007f3c3b488f7f in ray::core::CoreWorkerProcess::Shutdown() () from /tmp/ray/1642573567986/libcore_worker_library_java.so
  | #6  0x00007f3c98862a30 in ?? ()
  | #7  0x0000000000000001 in ?? ()
  | #8  0x00007f3cb7871528 in ?? ()
  | #9  0x0000000000000000 in ?? ()
```

**Comment by rkooo567** [comment]
Hmm that must be io_service thread
```
Thread 40 (Thread 0x7f3c28dfa700 (LWP 10406)):
#0  0x00007f3c3b69fc33 in ray::pubsub::Subscriber::SendCommandBatchIfPossible(ray::rpc::Address const&) () from /tmp/ray/1642573567986/libcore_worker_library_java.so
#1  0x00007f3c3b6a113b in std::_Function_handler<void (ray::Status const&, ray::rpc::PubsubCommandBatchReply const&), ray::pubsub::Subscriber::SendCommandBatchIfPossible(ray::rpc::Address const&)::{lambda(ray::Status, ray::rpc::PubsubCommandBatchReply const&)#1}>::_M_invoke(std::_Any_data const&, ray::Status const&, ray::rpc::PubsubCommandBatchReply const&) () from /tmp/ray/1642573567986/libcore_worker_library_java.so
#2  0x00007f3c3b5dff38 in std::_Function_handler<void (ray::Status const&, ray::rpc::GcsSubscriberCommandBatchReply const&), ray::gcs::(anonymous namespace)::GcsSubscriberClient::PubsubCommandBatch(ray::rpc::PubsubCommandBatchRequest const&, std::function<void (ray::Status const&, ray::rpc::PubsubCommandBatchReply const&)> const&)::{lambda(ray::Status const&, ray::rpc::GcsSubscriberCommandBatchReply const&)#1}>::_M_invoke(std::_Any_data const&, ray::Status const&, ray::rpc::GcsSubscriberCommandBatchReply const&) () from /tmp/ray/1642573567986/libcore_worker_library_java.so
#3  0x00007f3c3b5d6c92 in std::_Function_handler<void (ray::Status const&, ray::rpc::GcsSubscriberCommandBatchReply const&), ray::rpc::GcsRpcClient::GcsSubscriberCommandBatch(ray::rpc::GcsSubscriberCommandBatchRequest const&, std::function<void (ray::Status const&, ray::rpc::GcsSubscriberCommandBatchReply const&)> const&, long)::{lambda(ray::Status const&, ray::rpc::GcsSubscriberCommandBatchReply const&)#1}>::_M_invoke(std::_Any_data const&, ray::Status const&, ray::rpc::GcsSubscriberCommandBatchReply const&) () from /tmp/ray/1642573567986/libcore_worker_library_java.so
#4  0x00007f3c3b5d7009 in ray::rpc::ClientCallImpl<ray::rpc::GcsSubscriberCommandBatchReply>::OnReplyReceived() () from /tmp/ray/1642573567986/libcore_worker_library_java.so
#5  0x00007f3c3b3c583f in std::_Function_handler<void (), ray::rpc::ClientCallManager::PollEventsFromCompletionQueue(int)::{lambda()#1}>::_M_invoke(std::_Any_data const&) () from /tmp/ray/1642573567986/libcore_worker_library_java.so
#6  0x00007f3c3b6ee72f in EventTracker::RecordExecution(std::function<void ()> const&, std::shared_ptr<StatsHandle>) () from /tmp/ray/1642573567986/libcore_worker_library_java.so
#7  0x00007f3c3b6c2c9f in std::_Function_handler<void (), instrumented_io_context::post(std::function<void ()>, std::shared_ptr<StatsHandle>)::{lambda()#1}>::_M_invoke(std::_Any_data const&) () from /tmp/ray/1642573567986/libcore_worker_library_java.so
#8  0x00007f3c3b6c320c in boost::asio::detail::completion_handler<std::function<void ()>, boost::asio::io_context::basic_executor_type<std::allocator<void>, 0u> >::do_complete(void*, boost::asio::detail::scheduler_operation*, boost::system::error_code const&, unsigned long) () from /tmp/ray/1642573567986/libcore_worker_library_java.so
#9  0x00007f3c3bbf84bc in boost::asio::detail::scheduler::do_run_one(boost::asio::detail::conditionally_enabled_mutex::scoped_lock&, boost::asio::detail::scheduler_thread_info&, boost::system::error_code const&) () from /tmp/ray/1642573567986/libcore_worker_library_java.so
#10 0x00007f3c3bbf93f9 in boost::asio::detail::scheduler::run(boost::system::error_code&) () from /tmp/ray/1642573567986/libcore_worker_library_java.so
#11 0x00007f3c3bbf99d4 in boost::asio::io_context::run() () from /tmp/ray/1642573567986/libcore_worker_library_java.so
#12 0x00007f3c3b3ebf19 in ray::core::CoreWorker::RunIOService() () from /tmp/ray/1642573567986/libcore_worker_library_java.so
#13 0x00007f3cb7ab4de4 in ?? () from /lib/x86_64-linux-gnu/libstdc++.so.6
#14 0x00007f3cb8e7d609 in start_thread (arg=<optimized out>) at pthread_create.c:477
#15 0x00007f3cb8fdb293 in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95
```

This is weird though given this test doesn't change anything in shutdown path. cc @mwtian do you have any clue?

**Comment by rkooo567** [comment]
Hmm wonder if there's a deadlock condition here..

**Comment by mwtian** [comment]
> Hmm that must be io_service thread
> 
> ```
> Thread 40 (Thread 0x7f3c28dfa700 (LWP 10406)):
> #0  0x00007f3c3b69fc33 in ray::pubsub::Subscriber::SendCommandBatchIfPossible(ray::rpc::Address const&) () from /tmp/ray/1642573567986/libcore_worker_library_java.so
> #1  0x00007f3c3b6a113b in std::_Function_handler<void (ray::Status const&, ray::rpc::PubsubCommandBatchReply const&), ray::pubsub::Subscriber::SendCommandBatchIfPossible(ray::rpc::Address const&)::{lambda(ray::Status, ray::rpc::PubsubCommandBatchReply const&)#1}>::_M_invoke(std::_Any_data const&, ray::Status const&, ray::rpc::PubsubCommandBatchReply const&) () from /tmp/ray/1642573567986/libcore_worker_library_java.so
> #2  0x00007f3c3b5dff38 in std::_Function_handler<void (ray::Status const&, ray::rpc::GcsSubscriberCommandBatchReply const&), ray::gcs::(anonymous namespace)::GcsSubscriberClient::PubsubCommandBatch(ray::rpc::PubsubCommandBatchRequest const&, std::function<void (ray::Status const&, ray::rpc::PubsubCommandBatchReply const&)> const&)::{lambda(ray::Status const&, ray::rpc::GcsSubscriberCommandBatchReply const&)#1}>::_M_invoke(std::_Any_data const&, ray::Status const&, ray::rpc::GcsSubscriberCommandBatchReply const&) () from /tmp/ray/1642573567986/libcore_worker_library_java.so
> #3  0x00007f3c3b5d6c92 in std::_Function_handler<void (ray::Status const&, ray::rpc::GcsSubscriberCommandBatchReply const&), ray::rpc::GcsRpcClient::GcsSubscriberCommandBatch(ray::rpc::GcsSubscriberCommandBatchRequest const&, std::function<void (ray::Status const&, ray::rpc::GcsSubscriberCommandBatchReply const&)> const&, long)::{lambda(ray::Status const&, ray::rpc::GcsSubscriberCommandBatchReply const&)#1}>::_M_invoke(std::_Any_data const&, ray::Status const&, ray::rpc::GcsSubscriberCommandBatchReply const&) () from /tmp/ray/1642573567986/libcore_worker_library_java.so
> #4  0x00007f3c3b5d7009 in ray::rpc::ClientCallImpl<ray::rpc::GcsSubscriberCommandBatchReply>::OnReplyReceived() () from /tmp/ray/1642573567986/libcore_worker_library_java.so
> #5  0x00007f3c3b3c583f in std::_Function_handler<void (), ray::rpc::ClientCallManager::PollEventsFromCompletionQueue(int)::{lambda()#1}>::_M_invoke(std::_Any_data const&) () from /tmp/ray/1642573567986/libcore_worker_library_java.so
> #6  0x00007f3c3b6ee72f in EventTracker::RecordExecution(std::function<void ()> const&, std::shared_ptr<StatsHandle>) () from /tmp/ray/1642573567986/libcore_worker_library_java.so
> #7  0x00007f3c3b6c2c9f in std::_Function_handler<void (), instrumented_io_context::post(std::function<void ()>, std::shared_ptr<StatsHandle>)::{lambda()#1}>::_M_invoke(std::_Any_data const&) () from /tmp/ray/1642573567986/libcore_worker_library_java.so
> #8  0x00007f3c3b6c320c in boost::asio::detail::completion_handler<std::function<void ()>, boost::asio::io_context::basic_executor_type<std::allocator<void>, 0u> >::do_complete(void*, boost::asio::detail::scheduler_operation*, boost::system::error_code const&, unsigned long) () from /tmp/ray/1642573567986/libcore_worker_library_java.so
> #9  0x00007f3c3bbf84bc in boost::asio::detail::scheduler::do_run_one(boost::asio::detail::conditionally_enabled_mutex::scoped_lock&, boost::asio::detail::scheduler_thread_info&, boost::system::error_code const&) () from /tmp/ray/1642573567986/libcore_worker_library_java.so
> #10 0x00007f3c3bbf93f9 in boost::asio::detail::scheduler::run(boost::system::error_code&) () from /tmp/ray/1642573567986/libcore_worker_library_java.so
> #11 0x00007f3c3bbf99d4 in boost::asio::io_context::run() () from /tmp/ray/1642573567986/libcore_worker_library_java.so
> #12 0x00007f3c3b3ebf19 in ray::core::CoreWorker::RunIOService() () from /tmp/ray/1642573567986/libcore_worker_library_java.so
> #13 0x00007f3cb7ab4de4 in ?? () from /lib/x86_64-linux-gnu/libstdc++.so.6
> #14 0x00007f3cb8e7d609 in start_thread (arg=<optimized out>) at pthread_create.c:477
> #15 0x00007f3cb8fdb293 in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95
> ```
> 
> This is weird though given this test doesn't change anything in shutdown path. cc @mwtian do you have any clue?

Core worker during shutdown has issues like data races and use after free. I won't be surprised when subtle changes result in deadlock / seeing inconsistent state. We can try to check in #21593 first and see if the problem still exists.

**Comment by rkooo567** [comment]
@mwtian looks like this is unrelated to this PR. I found the same error from test-only change; https://buildkite.com/ray-project/ray-builders-pr/builds/22700#98c62403-3b53-4698-9477-5effb52d7701

**Review by scv119 — APPROVED** [review]

**Comment by rkooo567** [comment]
Hey @richardliaw do you mind reviewing the error messages here? https://github.com/ray-project/ray/pull/20903#issuecomment-1014230208 (feedback on formatting or messages itself)

**Comment by rkooo567** [comment]
@ericl seems like you removed the assignment? I will just go ahead and merge it in this case?

**Comment by rkooo567** [comment]
@wuisawesome do you mind giving me approval on .proto change?

**Review by wuisawesome — APPROVED** [review]

**Comment by rkooo567** [comment]
Java failure unrelated. This is happening in the master too
